File: Java/Trees/1339 MaximumProductOfSplittedBinaryTree.java
// LeetCode 1339. Maximum Product of Splitted Binary Tree
// Time Complexity: O(n)
// Space Complexity: O(n)

import java.util.LinkedList;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    private static final int MOD = 1_000_000_007;

    // --------------------------------------------------
    // Way 1: Two DFS passes
    // First DFS -> compute total sum
    // Second DFS -> compute max product
    // --------------------------------------------------
    public int maxProduct(TreeNode root) {
        long[] maxProd = new long[1];

        long totalSum = dfsSum(root);
        dfsProduct(root, totalSum, maxProd);

        return (int) (maxProd[0] % MOD);
    }

    private long dfsSum(TreeNode root) {
        if (root == null) return 0;
        return root.val + dfsSum(root.left) + dfsSum(root.right);
    }

    private long dfsProduct(TreeNode root, long totalSum, long[] maxProd) {
        if (root == null) return 0;

        long leftSum = dfsProduct(root.left, totalSum, maxProd);
        long rightSum = dfsProduct(root.right, totalSum, maxProd);

        long currSum = root.val + leftSum + rightSum;
        maxProd[0] = Math.max(maxProd[0], currSum * (totalSum - currSum));

        return currSum;
    }

    // --------------------------------------------------
    // Way 2: BFS for total sum + DFS for product
    // --------------------------------------------------
    public int maxProductBFS(TreeNode root) {
        long totalSum = 0;
        long[] maxProd = new long[1];

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        // BFS to compute total sum
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            totalSum += node.val;

            if (node.left != null) queue.offer(node.left);
            if (node.right != null) queue.offer(node.right);
        }

        dfsProduct(root, totalSum, maxProd);
        return (int) (maxProd[0] % MOD);
    }
}

File: C++/Trees/1339 MaximumProductOfSplittedBinaryTree.cpp
// LeetCode 1339. Maximum Product of Splitted Binary Tree
// Time Complexity: O(n)
// Space Complexity: O(n)

#include <queue>
#include <climits>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};

class Solution {
    static const int MOD = 1'000'000'007;

public:
    // --------------------------------------------------
    // Way 1: Two DFS passes
    // --------------------------------------------------
    int maxProduct(TreeNode* root) {
        long long maxProd = 0;

        long long totalSum = dfsSum(root);
        dfsProduct(root, totalSum, maxProd);

        return (int)(maxProd % MOD);
    }

private:
    long long dfsSum(TreeNode* root) {
        if (!root) return 0;
        return root->val + dfsSum(root->left) + dfsSum(root->right);
    }

    long long dfsProduct(TreeNode* root, long long totalSum, long long& maxProd) {
        if (!root) return 0;

        long long leftSum = dfsProduct(root->left, totalSum, maxProd);
        long long rightSum = dfsProduct(root->right, totalSum, maxProd);

        long long currSum = root->val + leftSum + rightSum;
        maxProd = max(maxProd, currSum * (totalSum - currSum));

        return currSum;
    }

public:
    // --------------------------------------------------
    // Way 2: BFS + DFS
    // --------------------------------------------------
    int maxProductBFS(TreeNode* root) {
        long long totalSum = 0;
        long long maxProd = 0;

        queue<TreeNode*> q;
        q.push(root);

        // BFS to compute total sum
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            totalSum += node->val;

            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }

        dfsProduct(root, totalSum, maxProd);
        return (int)(maxProd % MOD);
    }
};
